core/search/sort.hpp

```diff
@@ -58,85 +63,101 @@ struct score_ctx {
+    // write byte size instead of count to avoid multiply in DefaultScore call
+    return {absl::bit_cast<score_ctx*>(sizeof(score_t) * count), &DefaultScore,
+            &Noop};
   }
 
   // Returns scoring function setting `size` score buckets to `value`.
-  static ScoreFunction Constant(score_t value, uint32_t size) noexcept;
+  static ScoreFunction Constant(score_t value, uint32_t count) noexcept;
 
   // Returns scoring function setting a single score bucket to `value`.
   static ScoreFunction Constant(score_t value) noexcept;
 
   // Returns invalid scoring function.
-  static ScoreFunction Invalid() noexcept { return {nullptr, nullptr}; }
-
-  ScoreFunction() noexcept;
-  ScoreFunction(memory::managed_ptr<score_ctx>&& ctx,
-                const score_f func) noexcept
-    : ctx_(std::move(ctx)), func_(func) {}
-  ScoreFunction(std::unique_ptr<score_ctx>&& ctx, const score_f func) noexcept
-    : ScoreFunction(memory::to_managed<score_ctx>(std::move(ctx)), func) {}
-  ScoreFunction(score_ctx* ctx, const score_f func) noexcept
-    : ScoreFunction(memory::to_managed<score_ctx, false>(std::move(ctx)),
-                    func) {}
-  ScoreFunction(ScoreFunction&& rhs) noexcept;
-  ScoreFunction& operator=(ScoreFunction&& rhs) noexcept;
-
-  [[nodiscard]] score_ctx* Ctx() const noexcept { return ctx_.get(); }
-  [[nodiscard]] score_f Func() const noexcept { return func_; }
-
-  void Reset(memory::managed_ptr<score_ctx>&& ctx,
-             const score_f func) noexcept {
-    ctx_ = std::move(ctx);
-    func_ = func;
+  static ScoreFunction Invalid() noexcept { return {nullptr, nullptr, &Noop}; }
+
+  template<typename T, typename... Args>
+  static auto Make(score_f score, Args&&... args) {
+    return ScoreFunction{
+      new T{std::forward<Args>(args)...}, score,
+      [](score_ctx* ctx) noexcept { delete static_cast<T*>(ctx); }};
   }
 
-  void Reset(std::unique_ptr<score_ctx>&& ctx, const score_f func) noexcept {
-    ctx_ = memory::to_managed<score_ctx>(std::move(ctx));
-    func_ = func;
+  ScoreFunction() noexcept = default;
+  ScoreFunction(score_ctx& ctx, score_f score) noexcept
+    : ScoreFunction{&ctx, score, &Noop} {}
+  ScoreFunction(ScoreFunction&& rhs) noexcept
+    : ScoreFunction{std::exchange(rhs.ctx_, nullptr),
+                    std::exchange(rhs.score_, &DefaultScore),
+                    std::exchange(rhs.deleter_, &Noop)} {}
+  ScoreFunction& operator=(ScoreFunction&& rhs) noexcept {
+    if (this != &rhs) {
+      std::swap(ctx_, rhs.ctx_);
+      std::swap(score_, rhs.score_);
+      std::swap(deleter_, rhs.deleter_);
+    }
+    return *this;
   }
+  ~ScoreFunction() noexcept { deleter_(ctx_); }
 
-  void Reset(score_ctx* ctx, const score_f func) noexcept {
-    ctx_ = memory::to_managed<score_ctx, false>(ctx);
-    func_ = func;
+  void Reset(score_ctx& ctx, score_f score) noexcept {
+    deleter_(ctx_);
+    ctx_ = &ctx;
+    score_ = score;
+    deleter_ = &Noop;
```

Comment:
Do you need some safeguard such as 
```
if (&ctx != ctx_) {
  deleter_(ctx_);
  ctx_ = &ctx;
  score_ = score;
  deleter_ = &Noop;
}
```
here?

Reply:
Good catch!

Reply:
Hmm, not really
`IRS_ASSERT(&ctx != ctx_ || deleter_ == &Noop);`

Reply:
I will add this assert